translator/trace/oc_to_otlp.go

```diff
@@ -0,0 +1,433 @@
+
+	if len(td.Spans) != 0 {
+		resourceSpans.Spans = make([]*otlptrace.Span, 0, len(td.Spans))
+
+		for _, ocSpan := range td.Spans {
+			if ocSpan == nil {
+				// Skip nil spans.
+				continue
+			}
+
+			otlpSpan := ocSpanToOtlp(ocSpan)
+
+			if ocSpan.Resource != nil {
+				// Add a separate ResourceSpans item just for this span since it
+				// has a different Resource.
+				separateRS := &otlptrace.ResourceSpans{
+					Resource: ocNodeResourceToOtlp(nil, ocSpan.Resource),
+					Spans:    []*otlptrace.Span{otlpSpan},
+				}
+				resourceSpanList = append(resourceSpanList, separateRS)
+			} else {
+				// Otherwise add the span to the first ResourceSpans item.
+				resourceSpans.Spans = append(resourceSpans.Spans, otlpSpan)
+			}
+		}
+	}
+
+	return consumerdata.NewOTLPTraceData(resourceSpanList)
+}
+
+func timestampToUnixnano(ts *timestamp.Timestamp) uint64 {
+	return uint64(internal.TimestampToTime(ts).UnixNano())
+}
+
+func ocSpanToOtlp(ocSpan *octrace.Span) *otlptrace.Span {
+	attrs, droppedAttrCount := ocAttrsToOtlp(ocSpan.Attributes)
+	events, droppedEventCount := ocEventsToOtlp(ocSpan.TimeEvents)
+	links, droppedLinkCount := ocLinksToOtlp(ocSpan.Links)
+
+	childSpanCount := int32(0)
+	if ocSpan.ChildSpanCount != nil {
+		childSpanCount = int32(ocSpan.ChildSpanCount.Value)
+	}
+
+	otlpSpan := &otlptrace.Span{
+		TraceId:                ocSpan.TraceId,
+		SpanId:                 ocSpan.SpanId,
+		Tracestate:             ocTraceStateToOtlp(ocSpan.Tracestate),
+		ParentSpanId:           ocSpan.ParentSpanId,
+		Name:                   truncableStringToStr(ocSpan.Name),
+		Kind:                   ocSpanKindToOtlp(ocSpan.Kind, ocSpan.Attributes),
+		StartTimeUnixnano:      timestampToUnixnano(ocSpan.StartTime),
+		EndTimeUnixnano:        timestampToUnixnano(ocSpan.EndTime),
+		Attributes:             attrs,
+		DroppedAttributesCount: droppedAttrCount,
+		Events:                 events,
+		DroppedEventsCount:     droppedEventCount,
+		Links:                  links,
+		DroppedLinksCount:      droppedLinkCount,
+		Status:                 ocStatusToOtlp(ocSpan.Status),
+		LocalChildSpanCount:    childSpanCount,
+	}
+
+	return otlpSpan
+}
+
+func ocStatusToOtlp(ocStatus *octrace.Status) *otlptrace.Status {
+	if ocStatus == nil {
+		return nil
+	}
+	return &otlptrace.Status{
+		Code:    otlptrace.Status_StatusCode(ocStatus.Code),
+		Message: ocStatus.Message,
+	}
+}
+
+// Convert tracestate to W3C format. See the https://w3c.github.io/trace-context/
+func ocTraceStateToOtlp(ocTracestate *octrace.Span_Tracestate) string {
+	if ocTracestate == nil {
+		return ""
+	}
+	var sb strings.Builder
+	for i, entry := range ocTracestate.Entries {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		sb.WriteString(strings.Join([]string{entry.Key, entry.Value}, "="))
+	}
+	return sb.String()
+}
+
+func ocAttrsToOtlp(ocAttrs *octrace.Span_Attributes) (otlpAttrs []*otlpcommon.AttributeKeyValue, droppedCount uint32) {
+	if ocAttrs == nil {
+		otlpAttrs = nil
```

Comment:
nit: with the named returns I think these attributions here are not idiomatic

Reply:
I am not sure I understand. Do you mean that I can omit the assignments since return values are zero-initialized anyway?

Reply:
yes, that is correct.

Reply:
Agree. As per Effective Go:
>When named, they are initialized to the zero values for their types when the function begins; if the function executes a return statement with no arguments, the current values of the result parameters are used as the returned values. 

I wanted to show the return values explicitly, but my reading of the above paragraph is that it is idiomatic to rely on zero initialization.